### PR TITLE
Fix for memory widget causing crash

### PR DESCRIPTION
**Pull request by b10011**
The memory widget caused a crash of the WM.

### CODE REVIEW

**Comment by b10011** [comment]
Whoops, now the fix for non-graph CPU widget is also in the same pull req.

File: libqtile/widget/memory.py [hunk 1]
```diff
@@ -39,27 +39,33 @@ class Memory(base.ThreadedPollText):
     SwapFree: Returns amount of swap free
     SwapUsed: Returns amount of swap in use
 """
+
     orientations = base.ORIENTATION_HORIZONTAL
     defaults = [
-        ("fmt", "{MemUsed}M/{MemTotal}M", "Formatting for field names.")
+        ("format", "{MemUsed}M/{MemTotal}M", "Formatting for field names."),
```
Comment (pawamoy) [inline libqtile/widget/memory.py]: I don't think you should rename it `format`. It can shadow the built-in format function, making its use confusing. All other widgets also use `fmt`, this would break consistency. Wouldn't it be a breaking change as well, since users would need to update their code?

**Comment by pawamoy** [comment]
CPU widget, nice!! :rocket: 

Just added comments about `fmt` versus `format` :slightly_smiling_face:

File: libqtile/widget/cpu.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import psutil
+
+
+class CPU(base.ThreadedPollText):
+    orientations = base.ORIENTATION_HORIZONTAL
+
+    defaults = [
+        ("update_interval", 1.0, "Update interval for the CPU widget"),
+        (
+            "format",
```
Comment (pawamoy) [inline libqtile/widget/cpu.py]: Same remark as for the memory widget.

**Comment by b10011** [comment]
@pawamoy You can try it yourself, change the `format` to `fmt`, and it crashes again. There were two problems: usage of `fmt` and how it calls the inherited class.

The reason it crashes is because `libqtile.widget.base._TextBox` has property method `formatted_text` that tries to format the `fmt` string with `self._text`. That will result in `"{MemUsed}M/{MemTotal}M".format("N/A")` which will crash it. If we override that method with one that returns "hello", then our widget only says "hello" and can't be updated using update-method, because it uses `self.text`'s setter, and it calls `formatted_text`. `fmt` works for adding prefixes and postfixes, but it breaks for multiple variables. We could use `self._text` for storing tuple of numbers (cpu/memory data) but that would look really hacky and weird.

And clock-widget also uses `format` instead of `fmt`.

I could add `fmt` as a flag for the constructor, but store the format in `format`. That way it doesn't crash, and it doesn't break previous configs.

File: libqtile/widget/memory.py [hunk 3]
```diff
@@ -39,27 +39,33 @@ class Memory(base.ThreadedPollText):
...
+        ("update_interval", 1.0, "Update interval for the Memory"),
     ]
 
     def __init__(self, **config):
-        super().__init__(**config)
+        base.ThreadedPollText.__init__(self, **config)
```
Comment (pawamoy) [inline libqtile/widget/memory.py]: What is the reason for this change? Isn't it semantically the same?

**Comment by pawamoy** [comment]
Hmmm that's interesting information. Maybe there's some logic to fix in `libqtile.widget.base._TextBox` then (not implying you should change your PR). Thanks for the details.

File: libqtile/widget/memory.py (hunk 3 above)
Comment (b10011) [reply libqtile/widget/memory.py]: It's not, because `super().__init__` refers to `_TextBox.__init__` (iirc) and we need to init via `ThreadedPollText.__init__` or it crashes. I tried both changes individually and both crashes the WM.

**Comment by pawamoy** [comment]
I take back what I said before about the consistent use of `fmt`: only the widgets gmail, memory,  and pomodoro are using `fmt`, while all the other widgets use `format`. My bad :sweat_smile:!

**Comment by b10011** [comment]
@pawamoy No problem!

File: libqtile/widget/memory.py (hunk 3 above)
Comment (pawamoy) [reply libqtile/widget/memory.py]: Well, I'm a bit lost here.
- `ThreadedPollText` does not redefine `__init__`, so it must use its parent one, `InLoopPollText.__init__`, which itself calls `_TextBox.__init__`
- `super()` should do the same because it goes up one level, using the direct parent's `__init__` which is `ThreadedPollText.__init__`, etc.

But maybe there's things happening in `super()` that I don't know.

Anyway, what do you mean by both changes crash the WM? Which changes?

Comment (b10011) [reply libqtile/widget/memory.py]: Me too, it raised some error message earlier but now it worked just fine. I added a commit to change it back to `super().__init__`

I meant that changing `format` back to `fmt` crashes and using `super().__init__` crashes bot individually. But I was wrong about the super.

Comment (pawamoy) [reply libqtile/widget/memory.py]: OK thanks for the explanation :slightly_smiling_face:

**Comment by b10011** [comment]
I was checking other issues and noticed that this should also fix #1435

File: libqtile/widget/memory.py [hunk 4]
```diff
@@ -39,27 +39,33 @@ class Memory(base.ThreadedPollText):
...
+        ("update_interval", 1.0, "Update interval for the Memory"),
     ]
 
     def __init__(self, **config):
         super().__init__(**config)
         self.add_defaults(Memory.defaults)
 
+    def tick(self):
+        self.update(self.poll())
+        return self.update_interval
```
Comment (flacjacket) [inline libqtile/widget/memory.py]: What's the rationale for adding this? This seems to be changing the widget to behave like the in-loop poll text base class.

File: libqtile/widget/memory.py (hunk 1 above)
Comment (flacjacket) [reply libqtile/widget/memory.py]: Hmm, this is interesting. If the old method was broken, doing an api breaking change should be fine to fix it.

File: libqtile/widget/memory.py (hunk 4 above)
Comment (b10011) [reply libqtile/widget/memory.py]: If I remember correctly, without that it updated really randomly, occasionally it took a bit over second, sometimes over 3 seconds. With that it takes more consistently a bit over 1 second to update.

**Comment by b10011** [comment]
@flacjacket The structure, in general, was looking good but the inconsistent usage of the classes was causing problems.
I'm making custom prompt for myself with support for using xdg desktop entries and I noticed the Prompt is also quite cluttered and should support multiple completers at the same time.
Most of the widgets are quite small and I'd be willing to re-code/fix a bunch of them if there will be a redesign for base class structure and guidelines that must be followed to get the widgets merged.

File: libqtile/widget/memory.py (hunk 4 above)
Comment (flacjacket) [reply libqtile/widget/memory.py]: I think the reason that is the case is that it will run the update in the main thread, if that is desired it should base class the in-loop poll text widget, rather than subclassing the threaded poll text widget, but only if we know the function calls in the poll function are fast and not going to block.

File: libqtile/widget/memory.py (hunk 1 above)
Comment (ramnes) [reply libqtile/widget/memory.py]: It might be broken since the implementation of `fmt` in the base class. I'd be in favor of moving it to `format` so that it's consistent with other widgets, `fmt` being the "overall" format of the output and `format` being the internal formatting of the widget.

**Comment by flacjacket** [comment]
These changes seem good, thanks for the fixes! Sorry for the delay!
